applications/asset_tracker_v2/src/ext_sensors/ext_sensors.c

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <string.h>
+#include <drivers/sensor.h>
+#include "ext_sensors.h"
+#include <stdlib.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ext_sensors, CONFIG_EXTERNAL_SENSORS_LOG_LEVEL);
+
+/* Convert to s/m2 depending on the maximum measured range used for adxl362. */
+#if defined(CONFIG_ADXL362_ACCEL_RANGE_2G)
+#define ADXL362_RANGE_MAX_M_S2 19.6133
+#elif defined(CONFIG_ADXL362_ACCEL_RANGE_4G)
+#define ADXL362_RANGE_MAX_M_S2 39.2266
+#elif defined(CONFIG_ADXL362_ACCEL_RANGE_8G)
+#define ADXL362_RANGE_MAX_M_S2 78.4532
+#endif
+
+#define ADXL362_THRESHOLD_RESOLUTION_DECIMAL_MAX 2048
+
+/* Local accelerometer threshold value. Used to filter out unwanted values in
+ * the callback from the accelerometer.
+ */
+double threshold = ADXL362_RANGE_MAX_M_S2;
+
+struct env_sensor {
+	enum sensor_channel channel;
+	uint8_t *dev_name;
+	const struct device *dev;
+	struct k_spinlock lock;
+};
+
+static struct env_sensor temp_sensor = {
+	.channel = SENSOR_CHAN_AMBIENT_TEMP,
+	.dev_name = CONFIG_MULTISENSOR_DEV_NAME
+};
+
+static struct env_sensor humid_sensor = {
+	.channel = SENSOR_CHAN_HUMIDITY,
+	.dev_name = CONFIG_MULTISENSOR_DEV_NAME
+};
+
+static struct env_sensor accel_sensor = {
+	.channel = SENSOR_CHAN_ACCEL_XYZ,
+	.dev_name = CONFIG_ACCELEROMETER_DEV_NAME
```

Comment:
These dev_name values are all being initialized from string literals and therefore are in ROM, so there is no need to log_strdup() them in ext_sensors_mov_thres_set().

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/reference/logging/index.html#logging-strings

As an (even better) alternative, you may be able to avoid saving device names at all by just using DEVICE_DT_GET() and dropping all of these DEV_NAME config options. This sort of thing is deprecated anyway: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/zephyr/guides/kconfig/tips.html#options-that-specify-a-device-in-the-system-by-name

Reply:
I have removed the `log_strdup()`s from these and will look into fetching the device name from the device tree instead.

Reply:
My device tree skills are quite limited, so I'm not entirely sure how to go about this. Is this in the right direction?
https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/f16fff1931955f28eb60310121caa04ed18b9a16

Reply:
I'm using aliases so that other boards can more easily plug into the same code with a DT overlay

Reply:
> My device tree skills are quite limited, so I'm not entirely sure how to go about this. Is this in the right direction?
> [jtguggedal@f16fff1](https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/f16fff1931955f28eb60310121caa04ed18b9a16)
>
> I'm using aliases so that other boards can more easily plug into the same code with a DT overlay

Using aliases is fine, but you don't need to store the device label property at all anymore. You can just get the device straight away using DEVICE_DT_GET.

For example:

```c
struct env_sensor {
	enum sensor_channel channel;
	const struct device *dev;
	struct k_spinlock lock;
};

static struct env_sensor temp_sensor = {
	.channel = SENSOR_CHAN_AMBIENT_TEMP,
	.dev = DEVICE_DT_GET(DT_ALIAS(temp_sensor)),
};
```

Then you just have to check that `device_is_ready(temp_sensor.dev)` returns 0 in your init routine, and error out if it's not.

This has the following advantages:

- runs faster: no looping over devices in device_get_binding()
- fails faster: a missing device due to a bad devicetree configuration is a build error, not a runtime error
- smaller footprint: no dev_name

The name of the device is still accessible as `temp_sensor.dev->name` for debugging purposes.

Reply:
Oh, so that's how it works, pretty neat! I updated per your suggestions above: https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148

The alias stuff in `thingy91_nrf9160ns.overlay` felt a bit hacky to do this way, but it was the only way I found that worked. Is there a better way to alias a device on a given bus, like the BME680 or the ADXL362?

Reply:
> Thanks for explaining, these new DT macros offer some nice functionality that I wasn't aware of.

I strongly agree. Hat tip to @pabigot for making this happen for zephyr 2.5.

> Is there a better way to alias a device on a given bus, like the BME680 or the ADXL362

Depends on how you'd like to go about it. DEVICE_DT_GET() takes a devicetree node identifier, and there are a lot of ways to specify those.

I'm guessing you don't want to force the user to pick a BME680 for ambient temperature; e.g. if they have another sensor device that supplies SENSOR_CHAN_AMBIENT_TEMP, they should be able to use it easily by just tweaking their devicetree. And similarly for ADXL362.

Is that right? If so, you basically are limited either aliases or node labels.

Aliases you've already got a solution for.

If you want to use node labels, that'd look like this in the overlay:

```
&i2c2 {
	asset_tracker_temp: bme680@76 { ... };
};
```

Then you could say `DEVICE_DT_GET(DT_NODELABEL(asset_tracker_temp))` in C.

You could technically also use the `/zephyr,user` node (see the DT guide for documentation), but in this situation I think `/aliases` would be more idiomatic.

Reply:
> I'm guessing you don't want to force the user to pick a BME680 for ambient temperature; e.g. if they have another sensor device that supplies SENSOR_CHAN_AMBIENT_TEMP, they should be able to use it easily by just tweaking their devicetree. And similarly for ADXL362.

Yes, that's what I want to do. Looks like what I have on https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148 is fairly reasonable, then, so I'll submit a PR with that. Thanks for helping out on this.

Reply:
>  Looks like what I have on [jtguggedal@00002bf](https://github.com/jtguggedal/fw-nrfconnect-nrf/commit/00002bf09031fbf6dbbd30bff061fa058067148) is fairly reasonable, then, so I'll submit a PR with that. Thanks for helping out on this.

Sounds good, and sure thing.